src/Microsoft.AspNetCore.Routing/EndpointNameEndpointFinder.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal class EndpointNameEndpointFinder : IEndpointFinder<string>
+    {
+        private readonly DataSourceDependentCache<Dictionary<string, Endpoint[]>> _cache;
+
+        public EndpointNameEndpointFinder(CompositeEndpointDataSource dataSource)
+        {
+            _cache = new DataSourceDependentCache<Dictionary<string, Endpoint[]>>(dataSource, Initialize);
+        }
+
+        // Internal for tests
+        internal Dictionary<string, Endpoint[]> Entries => _cache.EnsureInitialized();
+
+        public IEnumerable<Endpoint> FindEndpoints(string address)
+        {
+            if (address == null)
+            {
+                throw new ArgumentNullException(nameof(address));
+            }
+
+            var entries = Entries;
```

Comment:
Added a comment. I'm just capturing the current value of the cache in a local for clarity. This would be important if we needed to access the cache twice per operation (for example).